Framework/Algorithms/src/SANSSolidAngle.cpp

```diff
@@ -0,0 +1,184 @@
+                  "The name of the workspace to be created as the output of "
+                  "the algorithm.  A workspace of this name will be created "
+                  "and stored in the Analysis Data Service.");
+
+  std::vector<std::string> exp_options{"Normal", "Tube", "Wing"};
+  declareProperty("Type", "Normal",
+                  boost::make_shared<StringListValidator>(exp_options),
+                  "Select the method to calculate the Solid Angle.\n"
+                  "Normal: cos^3(theta); Tube: cos(alpha)*cos^3(theta); "
+                  "Wing: cos^3(alpha);");
+}
+
+/**
+ * Returns the angle between the sample-to-pixel vector and its
+ * projection on the X-Z plane.
+ */
+double getYTubeAngle(const SpectrumInfo &spectrumInfo, size_t index) {
+  const V3D samplePos = spectrumInfo.samplePosition();
+
+  // Get the vector from the sample position to the detector pixel
+  V3D sampleDetVec = spectrumInfo.position(index) - samplePos;
+
+  // Get the projection of that vector on the X-Z plane
+  V3D inPlane = V3D(sampleDetVec);
+  inPlane.setY(0.0);
+
+  // This is the angle between the sample-to-detector vector
+  // and its project on the X-Z plane.
+  return sampleDetVec.angle(inPlane);
+}
+
+/**
+ *Returns correct angular function for the specified type
+ */
+std::function<double(int64_t)>
+getAngularFunction(const SpectrumInfo &spectrumInfo, const std::string &type) {
+  if (type == "Normal") {
+    return [&spectrumInfo](int64_t histogramIndex) {
+      double cosTheta = std::cos(spectrumInfo.twoTheta(histogramIndex));
+      return cosTheta * cosTheta * cosTheta;
+    };
+  } else if (type == "Tube") {
+    return [&spectrumInfo](int64_t histogramIndex) {
+      double cosTheta = std::cos(spectrumInfo.twoTheta(histogramIndex));
+      double cosAlpha = std::cos(getYTubeAngle(spectrumInfo, histogramIndex));
+      return cosTheta * cosTheta * cosAlpha;
+    };
+  } else if (type == "Wing") {
+    return [&spectrumInfo](int64_t histogramIndex) {
+      double cosAlpha = std::cos(getYTubeAngle(spectrumInfo, histogramIndex));
+      return cosAlpha * cosAlpha * cosAlpha;
+    };
+  } else {
+    throw std::runtime_error("Invalid type of correction");
+  }
+}
+
+/**
+ * Compute the solid angle using the correct angular function
+ */
+class CalculateSolidAngle {
+public:
+  CalculateSolidAngle(const double pixelArea, const SpectrumInfo &spectrumInfo,
+                      const ComponentInfo &componentInfo,
+                      std::function<double(int64_t)> &&f)
+      : m_pixelArea(pixelArea), m_spectrumInfo(spectrumInfo),
+        m_componentInfo(componentInfo), m_f(std::move(f)) {}
+  double operator()(int64_t histogramIndex) const {
+    Kernel::V3D scaleFactor = m_componentInfo.scaleFactor(histogramIndex);
+    double scaledPixelArea = m_pixelArea * scaleFactor[0] * scaleFactor[1];
+    double angularTerm = m_f(histogramIndex);
+    double distance = m_spectrumInfo.l2(histogramIndex) * 1.e3; // mm
```

Comment:
I think it's this line that is wrong, and this can explain the large difference you are getting. When using the `cosˆ3` formalism, this distance should be the distance of the rectangular detector from the sample along the beam direction instead of the l2 of the current pixel. 
![Screen Shot 2019-06-14 at 12 47 41](https://user-images.githubusercontent.com/16865759/59504512-682b6a80-8ea3-11e9-8e06-814719042b2f.png)

Reply:
And here is the formula 22 from the Brulet paper:
![Screen Shot 2019-06-14 at 13 02 31](https://user-images.githubusercontent.com/16865759/59505133-d45a9e00-8ea4-11e9-8c0a-32f33020e6f8.png)

Reply:
Looking the requirements from our instrument scientists, it looks like you're right. What's the reference you're using? I don't recognize it.

Reply:
I was looking at `Small-Angle Neutron
Scattering and
Applications in Soft
Condensed Matter` by I. GRILLO